### PR TITLE
[Feature] Add Gumiho speculative decoding method

### PR DESCRIPTION
**Pull request by userljz**
## Purpose

Adds support for **Gumiho** as a new speculative decoding method
(`speculative_config={"method": "gumiho", ...}`).

Gumiho is a hybrid drafter:

1. The first two speculative tokens are generated autoregressively by a
   small EAGLE-style transformer head.
2. Every additional speculative token is generated **in parallel** by
   per-position MLP heads conditioned on the embeddings and hidden states
   of the first two draft tokens.

This trades a small amount of acceptance rate for a meaningful reduction in
drafter latency when `num_speculative_tokens > 2`, with no change to the
verifier or rejection sampler.

The drafter wraps a 2-layer Llama backbone, so it is implemented as an EAGLE
proposer subclass (`GumihoProposer`) plus a separate drafter model
(`GumihoLlamaForCausalLM`). All EAGLE-style infrastructure (padded drafter
batch, KV slot bookkeeping, cudagraph dispatch, attention metadata building,
rejection sampling) is reused as-is.

### Reference

- Paper: Li et al., [_Gumiho: A Hybrid Architecture to Prioritize Early
  Tokens in Speculative Decoding_](https://arxiv.org/abs/2503.10135),
  **ICML 2025** ([PMLR](https://proceedings.mlr.press/v267/li25dg.html)).
- Reference code: <https://github.com/AMD-AIG-AIMA/Gumiho>
- Pre-trained drafter: [`amd/Gumiho-llama3-8b`](https://huggingface.co/amd/Gumiho-llama3-8b)
- This PR is submitted by the original authors of the paper.

## Changes

**New files (5):**

| File | LoC | Purpose |
| --- | --- | --- |
| `vllm/v1/spec_decode/gumiho.py` | 74 | `GumihoProposer`, a thin `EagleProposer` subclass that overrides two hooks. |
| `vllm/model_executor/models/gumiho.py` | 437 | `GumihoLlamaForCausalLM` (transformer draft head + parallel MLP heads + LM head + weight loader). |
| `vllm/transformers_utils/configs/gumiho.py` | 80 | `GumihoConfig` HF config wrapper. |
| `tests/v1/spec_decode/test_gumiho.py` | 350 | CPU unit tests for config wrapping, MLP-head shape contract, hook behaviour, and `use_eagle()` wiring. |
| `docs/features/speculative_decoding/gumiho.md` | 72 | User-facing docs + example config. |

**Modified files (9):**

- `vllm/v1/spec_decode/llm_base_proposer.py` — adds two no-op hooks
  (`_init_draft_hidden_states_list`, `_maybe_get_mlp_draft_token_ids`) that
  the sequential drafting loop in `propose()` calls. Default behaviour is
  unchanged for all existing proposers.
- `vllm/v1/worker/gpu_model_runner.py` — wires up `GumihoProposer` and adds
  it to the drafter type union. Matched **before** `use_eagle()` so that
  Gumiho doesn't silently fall back to a plain `EagleProposer`.
- `vllm/v1/worker/gpu/spec_decode/__init__.py` — explicit `NotImplementedError`
  for Gumiho on the V2 GPU runner (it isn't supported there yet, so we surface
  a clear error rather than silently using `EagleSpeculator`).
- `vllm/config/speculative.py` — adds `gumiho` to `SpeculativeMethod`, to
  the `use_eagle()` set (so all EAGLE-style code paths apply), to the
  draft-TP=1 forced set, and adds the auto-detect / config-wrap branches.
- `vllm/model_executor/models/registry.py` — registers `Gumiho`,
  `GumihoModel`, and `GumihoLlamaForCausalLM` architectures.
- `vllm/transformers_utils/{config.py,configs/__init__.py}` — registers
  `GumihoConfig` and its lazy import path.
- `docs/features/speculative_decoding/README.md` — adds Gumiho to the
  methods list and to the comparison table.
- `examples/features/speculative_decoding/spec_decode_offline.py` — adds a
  `--method gumiho` branch with sane defaults.

Net diff: **110 added / 4 removed in modified files**, **1013 LoC across 5 new files**.

### Why a new proposer instead of extending `EagleProposer`?

`GumihoProposer` is a one-class subclass of `EagleProposer` that overrides
exactly two hooks (`_init_draft_hidden_states_list`,
`_maybe_get_mlp_draft_token_ids`) on the shared `SpecDecodeBaseProposer`.
Everything else — padded drafter batch, KV slot mapping, attention metadata,
cudagraph dispatch — is inherited unchanged. The hooks are no-ops in the
base class so this PR is non-functional for every other proposer.

### Where the MLP heads fit in

```
target model forward
    -> target hidden states
    -> GumihoProposer.propose() (inherits SpecDecodeBaseProposer.propose)
        -> transformer draft head: token 1   (sample_hidden_states)
        -> transformer draft head: token 2   (first loop iteration)
        -> _maybe_get_mlp_draft_token_ids:
             GumihoLlamaForCausalLM.generate_mlp_draft_token_ids
             -> remaining tokens generated in parallel by MLP heads
        -> break out of the sequential loop
    -> V1 target verify
    -> V1 rejection sampler
```

### Greedy-only MLP path

The MLP heads currently produce argmax token ids only, so when
`draft_sample_method=probabilistic` the proposer falls back to the regular
sequential drafting path. Extending the MLP heads to emit logits/probs is
left to a follow-up.

### Not in this PR (potential follow-ups)

- Tree attention / FTA verification (the released Gumiho training code has
  this; the V1 verifier does not support tree drafts yet).
- Multimodal drafters.
- The V2 GPU model runner path.

## Test Plan

1. `pre-commit run --files <changed files>` (ruff check + ruff format pass).
2. `pytest tests/v1/spec_decode/test_gumiho.py` — CPU unit tests for config,
   MLP shape contract, OOB-head filtering, and `GumihoProposer` hooks.
3. End-to-end smoke test on 1× MI300X with the published drafter:

   ```bash
   python examples/features/speculative_decoding/spec_decode_offline.py \
     --method gumiho \
     --model-dir meta-llama/Meta-Llama-3-8B-Instruct \
     --gumiho-dir amd/Gumiho-llama3-8b \
     --num-spec-tokens 3 \
     --dataset-name hf --dataset-path philschmid/mt-bench --num-prompts 80 \
     --temp 0 --top-p 1.0
   ```

### Unit tests

```
$ pytest tests/v1/spec_decode/test_gumiho.py -v
========================= 24 passed, 17 warnings in 7.95s =========================
```

All 24 unit tests pass on the V1 ROCm path (MI300X, vLLM v0.21.0 + patched
``llm_base_proposer.py`` hooks; the exact same code is what is being merged
on top of ``main``).

### End-to-end acceptance length

Setup: ``meta-llama/Meta-Llama-3-8B-Instruct`` target + ``amd/Gumiho-llama3-8b``
drafter, MI300X (1× GPU), ``temperature=0``, ``top_p=1.0``,
``output_len=256``, ``mt-bench`` 80 prompts, command:

```bash
python examples/features/speculative_decoding/spec_decode_offline.py \
  --method gumiho --gumiho-dir /cache/gumiho/amd-Gumiho-llama3-8b \
  --num-spec-tokens K \
  --dataset-name hf --dataset-path philschmid/mt-bench --num-prompts 80 \
  --temp 0 --top-p 1.0 --max-model-len 8192 --output-len 256
```

| Method | K | mean acceptance length | acc@pos0 | acc@pos1 | acc@pos2 | acc@pos3 | acc@pos4 | acc@pos5 |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| (no spec) | – | 1.00 | – | – | – | – | – | – |
| **gumiho** | 3 | **2.11** | 0.61 | 0.33 | 0.18 | – | – | – |
| **gumiho** | 6 | **2.26** | 0.60 | 0.32 | 0.17 | 0.09 | 0.05 | 0.03 |

`K=6` exercises 4 of the 5 MLP heads contained in the released
``amd/Gumiho-llama3-8b`` checkpoint and is closer to the configuration
reported in the paper. Acceptance lengths align with the paper's Llama-3
8B numbers (Table 1) within the differences expected from the V1 verifier
not yet using tree attention.

### Throughput (batched, 80 prompts, single MI300X)

| Configuration | output tokens/s | wall-clock |
| --- | --- | --- |
| no spec | 4682 | 3.6 s |
| gumiho, k=3 | 3026 | 5.6 s |
| gumiho, k=6 | 2907 | 5.9 s |

**Note**: ``mt-bench --num-prompts 80`` runs all 80 prompts as one fully
batched offline job that finishes in a few seconds, which is the
throughput-saturated regime — speculative decoding deliberately trades some
of that aggregate throughput for lower **per-request latency** at low
QPS / batch=1, where the speedups in the paper were measured. A
latency-focused benchmark (single request, ``benchmark_latency.py``) is a
better fit for comparing against EAGLE/EAGLE-2 here and is left for a
follow-up commit.

## Backward Compatibility

- All hook methods in `SpecDecodeBaseProposer` are no-ops by default. No
  existing proposer changes behaviour.
- `use_eagle()` returns True for `gumiho`, but the drafter branch in
  `gpu_model_runner.py` matches `method == "gumiho"` first, so existing
  EAGLE / EAGLE3 / MTP / DFlash paths are unaffected.
- V2 GPU model runner raises `NotImplementedError` for Gumiho with a clear
  message instead of silently falling back.

### CODE REVIEW

File: vllm/model_executor/models/gumiho.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+        mlp_input = torch.cat(mlp_inputs, dim=-1)
+
+        next_token_ids = []
+        for block in self.mlp[:num_tokens]:
+            hidden_states = block(mlp_input)
+            logits = self.compute_logits(hidden_states)
+            next_token_ids.append(logits.argmax(dim=-1))
+        if not next_token_ids:
+            return None
+        return torch.stack(next_token_ids, dim=1)
```
[the remark this answers is not in the document]
Comment (userljz) [reply vllm/model_executor/models/gumiho.py]: Good catch, thanks! Applied in 9a6e6939c. Verified equivalent on MI300X: mt-bench K=6 mean acceptance length 2.25 (was 2.26, within argmax tie-break noise).

**Comment by userljz** [comment]
Friendly ping — a few notes for reviewers:

* RFC discussion: #43545 (one-pager with motivation + design)
* Validated end-to-end on AMD MI300X (ROCm) with the published
  drafter `amd/Gumiho-llama3-8b`, mt-bench 80 prompts:
  mean acceptance length **2.11** (k=3) / **2.26** (k=6).
  Numbers align with the Llama-3 8B row of Table 1 in the paper
  (ICML 2025, arXiv:2503.10135).
* @gemini-code-assist's perf suggestion (batched LM-head projection
  in `generate_mlp_draft_token_ids`) is now addressed in a follow-up
  commit.
* The drafter touches only `vllm/v1/spec_decode/llm_base_proposer.py`
  via two no-op hooks; every other proposer is bit-equivalent to
  before this PR.

cc @tjtanaa, would appreciate a look if you have bandwidth.

**Review by benchislett — CHANGES_REQUESTED** [review]
I'm generally a fan of block-wise decoding for speculative-decoding as opposed to fully autoregressive or fully parallel. I've done some studies in the past indicating that 2-parallel + 9-parallel would be more effective than 11-parallel drafting. However, I'm skeptical of this implementation:

- Hardcoding the bound on num_speculative_tokens seems odd. A properly trained P-EAGLE drafter can be used for arbitrary block-decoding regimens. This would be more general I think. I'm not sure if the paper shows special benefit for training separate networks for first-few-tokens v.s. the rest of them.
- DFlash has shown particularly strong results for fully parallel drafting, can you compare performance against this method? If results are not compelling it will be hard to invest effort maintaining this extra technique.
